### PR TITLE
contrib: add first draft of agent configuration docs

### PR DESCRIPTION
**Pull request by dnephin**
Building on #10281

[Preview](https://github.com/hashicorp/consul/tree/dnephin/contrib-docs-config/contributing/config).

Adds some developer docs for agent configuration. There aren't many details about auto-config, but hopefully the links to the source are a reasonable start.

### CODE REVIEW

File: contributing/config/README.md [hunk 1]
```diff
@@ -1,3 +1,53 @@
 # Agent Configuration
 
-- [Checklist for adding a new field](./checklist-adding-config-fields.md)
+The [Agent Configuration] is the primary mechanism for configuring Consul. Agent
+Configuration also allows for specifying Config Entries, Services, and Checks that
```
Comment (eculver) [inline contributing/config/README.md:4]: would it be helpful to link out to the main docs on these topics for good measure (similar to what is there for Agent Configuration).

File: contributing/config/README.md [hunk 2]
```diff
@@ -1,3 +1,53 @@
...
+See also the [checklist for adding a new field] to the configuration.
+
+[hcl]: https://github.com/hashicorp/hcl/tree/hcl1
+[Agent Configuration]: https://www.consul.io/docs/agent/options
+[checklist for adding a new field]: ./checklist-adding-config-fields.md
+[Auto-Config]: #auto-config
+## Code
+
+The Agent Configuration is implemented in [agent/config], and the primary entrypoint is
+[Load]. Config loading is performced in phases:
```
Comment (eculver) [inline contributing/config/README.md:19]: `s/performced/performed`

**Review by eculver — APPROVED** [review]
Docs!

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/409635.

**Pull request merged**